Replace JavaScriptEvaluationResult's IPC format with a structured map

achristensen07 opened the pull request:
#### 83c2523fa0c911e3459629da626ea6d3e6c6d5fc
<pre>
Replace JavaScriptEvaluationResult&apos;s IPC format with a structured map
<a href="https://bugs.webkit.org/show_bug.cgi?id=288788">https://bugs.webkit.org/show_bug.cgi?id=288788</a>
<a href="https://rdar.apple.com/145804771">rdar://145804771</a>

Reviewed by Sihui Liu.

In order to maintain maximum compatibility, we continue to do a roundtrip
through SerializedScriptValue, but we do so now entirely inside the web content
process.  This preserves some strange behavior related to things like blob
serialization which we may want to revisit later, but for now we don&apos;t change
any observable behavior.  This is verified by some excellent tests that
I got from Simon Lewis and included in this PR from his work on a slightly
different approach in <a href="https://github.com/WebKit/WebKit/pull/40911">https://github.com/WebKit/WebKit/pull/40911</a>
Simon&apos;s tests are more extensive than anything we&apos;ve tested before, which is
why they hit several assertions I needed to remove.

The serialization format is a map of identifiers to values of different types
with a root identifier of where to start.  This allows us to preserve pointer
equality of dictionaries that contain each other and keys that are reused as
values and other interesting things.

I also adopt JavaScriptEvaluationResult for the places where the extensions
code deserializes a JavaScript evaluation result, such as
Messages::WebInspectorUIExtensionController::EvaluateScriptForExtension.

* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toWK const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::root const):
(WebKit::JavaScriptEvaluationResult::map const):
(WebKit::JavaScriptEvaluationResult::wireBytes const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toID const):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::convertToObjC):
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
(WebKit::CoreIPCNumber::numberHolderForNumber): Deleted.
(WebKit::CoreIPCNumber::CoreIPCNumber): Deleted.
(WebKit::CoreIPCNumber::createCFNumber const): Deleted.
(WebKit::CoreIPCNumber::get const): Deleted.
(WebKit::CoreIPCNumber::toID const): Deleted.
* Source/WebKit/Shared/cf/CoreIPCNumber.mm: Added.
(WebKit::CoreIPCNumber::numberHolderForNumber):
(WebKit::CoreIPCNumber::CoreIPCNumber):
(WebKit::CoreIPCNumber::createCFNumber const):
(WebKit::CoreIPCNumber::get const):
(WebKit::CoreIPCNumber::toID const):
* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SerializedScriptValue::deserializationContext):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(ReturnTypes)):

Canonical link: <a href="https://commits.webkit.org/291332@main">https://commits.webkit.org/291332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/711b738860ce140f9d40a72a36aab0fbf369d1a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92688 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12237 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1860 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43199 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94738 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20691 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28410 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95690 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83909 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1540 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42527 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85398 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1510 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99703 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91354 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19740 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79803 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79283 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23807 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14774 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19724 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114002 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19411 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32937 "Found 8 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.bytecode-cache, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-slow-memory, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-eager (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22871 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21152 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
```diff
@@ -26,13 +26,225 @@
+        return nil;
+
+    RetainPtr result = toID(it->value);
+    m_instantiatedObjects.set(root, result);
```
We should update `m_instantiatedObjects` before recursively converting its children, to avoid invoking `toID()` for the same Identifier.

szewai commented on Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
```diff
@@ -26,13 +26,225 @@
+
+    auto identifier = JSObjectID::generate();
+    m_map.add(identifier, toVariant(object));
+    m_objectsInMap.set(object, identifier);
```
I think we need to update `m_objectsInMap` before `m_map`, otherwise there could be infinite loop for object refers to itself. E.g. for `array` object in `var array = []; array.push(array);`:
1. `addObjectToMap` creates ID=1 for it, since it's not in `m_objectsInMap`
2. `toVariant` converts it by invoking `addObjectToMap` on the first element (so it's the same object)
3. `addObjectToMap` creates ID=2 for it, since it's not in `m_objectsInMap`...

We might want to add a test for this if there isn't one.

szewai reviewed (approved):
Nice! r=me if bots are happy

achristensen07 committed:
Replace JavaScriptEvaluationResult's IPC format with a structured map
https://bugs.webkit.org/show_bug.cgi?id=288788
rdar://145804771

Reviewed by Sihui Liu.

In order to maintain maximum compatibility, we continue to do a roundtrip
through SerializedScriptValue, but we do so now entirely inside the web content
process.  This preserves some strange behavior related to things like blob
serialization which we may want to revisit later, but for now we don't change
any observable behavior.  This is verified by some excellent tests that
I got from Simon Lewis and included in this PR from his work on a slightly
different approach in https://github.com/WebKit/WebKit/pull/40911
Simon's tests are more extensive than anything we've tested before, which is
why they hit several assertions I needed to remove.

The serialization format is a map of identifiers to values of different types
with a root identifier of where to start.  This allows us to preserve pointer
equality of dictionaries that contain each other and keys that are reused as
values and other interesting things.

I also adopt JavaScriptEvaluationResult for the places where the extensions
code deserializes a JavaScript evaluation result, such as
Messages::WebInspectorUIExtensionController::EvaluateScriptForExtension.

* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toWK const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::root const):
(WebKit::JavaScriptEvaluationResult::map const):
(WebKit::JavaScriptEvaluationResult::wireBytes const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toID const):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::convertToObjC):
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
(WebKit::CoreIPCNumber::numberHolderForNumber): Deleted.
(WebKit::CoreIPCNumber::CoreIPCNumber): Deleted.
(WebKit::CoreIPCNumber::createCFNumber const): Deleted.
(WebKit::CoreIPCNumber::get const): Deleted.
(WebKit::CoreIPCNumber::toID const): Deleted.
* Source/WebKit/Shared/cf/CoreIPCNumber.mm: Added.
(WebKit::CoreIPCNumber::numberHolderForNumber):
(WebKit::CoreIPCNumber::CoreIPCNumber):
(WebKit::CoreIPCNumber::createCFNumber const):
(WebKit::CoreIPCNumber::get const):
(WebKit::CoreIPCNumber::toID const):
* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SerializedScriptValue::deserializationContext):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(ReturnTypes)):

Canonical link: https://commits.webkit.org/291332@main
